### ISSUE TITLE
Recommended way to use Tensorboard in a pipelines run

### ISSUE THREAD

**Issue by mgiessing** [kubeflow/pipelines#7163]
Hi,
I'm currently wondering what is the recommended way to include Tensorboard **inside** a pipeline run.

For non-pipeline tb I followed a minimal example that trains mnist on a notebook server --> https://www.youtube.com/watch?v=eMDF2Bk8YRY which works good!

For pipelines I currently feel that there are two options (probably even more)

1. Use **tensorboard viewer**, therefore I need to upload the logs to MinIO (I'm using local Kubeflow deployment) similar to the following Notebook snippet. The "disadvantage" I see there is that I have to wait until the component is completed until I can start the TensorBoard.

```python

from typing import NamedTuple
def mnist_train(model_file: str) -> NamedTuple("EvaluationOutput", [('mlpipeline_ui_metadata', 'UI_metadata')]):

    from datetime import datetime
    import tensorflow as tf
    
    model = tf.keras.models.Sequential([
      tf.keras.layers.Flatten(input_shape=(28, 28)),
      tf.keras.layers.Dense(512, activation=tf.nn.relu),
      tf.keras.layers.Dropout(0.2),
      tf.keras.layers.Dense(10, activation=tf.nn.softmax)
    ])
    
    model.compile(optimizer='adam',
                  loss='sparse_categorical_crossentropy',
                  metrics=['accuracy'])
    
    print(model.summary())
    
    mnist = tf.keras.datasets.mnist
    (x_train, y_train),(x_test, y_test) = mnist.load_data()
    x_train, x_test = x_train / 255.0, x_test / 255.0

    callbacks = [
      tf.keras.callbacks.TensorBoard(log_dir='logs/' + datetime.now().date().__str__()),
      tf.keras.callbacks.EarlyStopping(patience=2, monitor='val_loss'),
    ]
    
    model.fit(x_train, y_train, batch_size=32, epochs=3, callbacks=callbacks, validation_data=(x_test, y_test))

    ##To Do: Upload the logs to MinIO and point the source path to MinIO instead of gs://...

    import json
    from collections import namedtuple

    # Exports a sample tensorboard:
    metadata = {
      'outputs' : [{
        'type': 'tensorboard',
        'source': 'gs://ml-pipeline-dataset/tensorboard-train',
      }]
    }
    
    out_tuple = namedtuple("EvaluationOutput", ["mlpipeline_ui_metadata"])
    
    return out_tuple(json.dumps(metadata))


model_train_op = comp.func_to_container_op(func=mnist_train, base_image="quay.io/ibm/kubeflow-component-tensorflow-cpu:latest")

@dsl.pipeline(
   name='Mnist pipeline',
   description='A toy pipeline that performs mnist model training.'
)
def mnist_pipeline(model_file: str = 'mnist_model.h5'):
    model_train_op(model_file=model_file)

arguments = {"model_file":"mnist_model.h5"}

# Submit pipeline directly from pipeline function
run_result = client.create_run_from_pipeline_func(mnist_pipeline, arguments=arguments, namespace='kubeflow-user-example-com')

```

2. Mount a volume (maybe the NB server volume?) into the pipeline component and start the Tensorboard via the Kubeflow Dashboard Tab (the way which is also shown in the yt video). Probably sth. similar to this example I guess: https://github.com/kubeflow/pipelines/blob/master/samples/core/volume_ops/volume_ops.py

The advantage I see with the latter is that I can also start using/inspecting TensorBoard while the model is still training, however I read, that those will make the pipeline less flexible/reusable and should be avoided if possible...?
Maybe someone can shed some light on the recommended way to achieve this.

Thanks!

**Comment by chensun** [comment]
Per our doc, your option 1 seems to be our recommendation: https://www.kubeflow.org/docs/components/pipelines/sdk/output-viewer/#tensorboard

**Comment by irabanillo91** [comment]
Have you been able to get the viewer linked to the component? I'm able to generate the artifact named "mlpipeline_ui_metadata" as explained in the docs
![image](https://user-images.githubusercontent.com/60658860/150397983-9e227cf7-f79d-4cac-b58e-970c5cac9599.png)

which looks like:
`{"outputs": [{"type": "tensorboard", "source": "s3://my_bucket/ai_kubeflow_example/pipeline/CatsDogs_inaki/9c735a49-f1b4-44ac-99b3-18d768e46600/training-component-v2/logs"}]}`

However, nothing shows in the "Visualizations" tab
![image](https://user-images.githubusercontent.com/60658860/150398479-cd476ddd-3d1a-41d6-911f-73c9581fbc41.png)

Alternatively, if I start a Tensorboard server pointing to that same S3 URI, all looks good.

**Comment by mgiessing** [comment]
Yes it does work for me. How does your function look like which you'd expect to link the viewer?

**Comment by irabanillo91** [comment]
I've tried to simplify it as much as possible. It looks like this:

```
from kfp.aws import use_aws_secret
from kfp.compiler import Compiler
from kfp.components import func_to_container_op
from kfp.dsl import PipelineExecutionMode, pipeline
from ai_kubeflow_example.kubeflow.images import DOCKER_IMAGE

from typing import NamedTuple
def simple_component()-> NamedTuple("EvaluationOutput", [('mlpipeline_ui_metadata', 'UI_metadata')]):
    import json
    # Exports a sample tensorboard:
    metadata = {
        'outputs': [{
            'type': 'tensorboard',
            'source': 'gs://ml-pipeline-dataset/tensorboard-train',
        }]
    }
    from collections import namedtuple
    out_tuple = namedtuple("EvaluationOutput", ["mlpipeline_ui_metadata"])

    return out_tuple(json.dumps(metadata))


@pipeline(name="Simple component", description="Simple component",)
def example_pipeline():
    simple_component_op = func_to_container_op(func=simple_component, base_image=DOCKER_IMAGE)
    simple_component_op().apply(use_aws_secret("mlpipeline-minio-artifact", "accesskey", "secretkey", "eu-west-2"))


if __name__ == "__main__":
    Compiler(mode=PipelineExecutionMode.V2_COMPATIBLE).compile(training_pipeline_v2, f"{__file__[:-len('.py')]}.yaml")
```

However, nothing shows up in the Viz tab
![image](https://user-images.githubusercontent.com/60658860/150535783-a1fa3ab3-05e9-44ac-a410-a3a503dd503d.png)

![image](https://user-images.githubusercontent.com/60658860/150535748-41c4e3e4-e976-4d23-9238-cf19f1e36e93.png)

**Comment by mgiessing** [comment]
I don't use AWS, nor do I use x86 but your code seems to link the Tensorboard for me if I adjust it minimally according to my environment:

```python
import kfp
from kfp.compiler import Compiler
from kfp.components import func_to_container_op
from kfp.dsl import PipelineExecutionMode, pipeline

from typing import NamedTuple
def simple_component()-> NamedTuple("EvaluationOutput", [('mlpipeline_ui_metadata', 'UI_metadata')]):
    import json
    # Exports a sample tensorboard:
    metadata = {
        'outputs': [{
            'type': 'tensorboard',
            'source': 'gs://ml-pipeline-dataset/tensorboard-train',
        }]
    }
    from collections import namedtuple
    out_tuple = namedtuple("EvaluationOutput", ["mlpipeline_ui_metadata"])

    return out_tuple(json.dumps(metadata))


@pipeline(name="Simple component", description="Simple component")
def example_pipeline():
    simple_component_op = func_to_container_op(func=simple_component, base_image="python:3.7")
    simple_component_op() #.apply(use_aws_secret("mlpipeline-minio-artifact", "accesskey", "secretkey", "eu-west-2"))


if __name__ == "__main__":
    kfp.Client().create_run_from_pipeline_func(example_pipeline, arguments={}, namespace='user-example-com')
    #Compiler(mode=PipelineExecutionMode.V2_COMPATIBLE).compile(training_pipeline_v2, f"{__file__[:-len('.py')]}.yaml")
```

![image](https://user-images.githubusercontent.com/40735330/150554748-26973841-b8ea-4794-aff6-a40e2271b08e.png)

I saw you first compile the function into a workflow.yaml file which you then use to create a run I guess? Shouldn't you use the correct pipeline func `example_pipeline` instead of `training_pipeline_v2` then...? However you seem to be able to get a correct run out.

**Comment by irabanillo91** [comment]
Oh, my bad, that was a copy&paste issue. In my code I generate the pipeline for `example_pipeline`, not `training_pipeline_v2`. Then I manually upload the yaml file and create the run from the dashboard.

Can I ask you what version of kubeflow you're using (we're on [1.4](https://www.kubeflow.org/docs/releases/kubeflow-1.4/))? And also what version of the python package kfp you're using (we're on `kfp=1.8.10`)?

**Comment by mgiessing** [comment]
Okay! I'm using Kubeflow v1.3.0 & kfp 1.8.5 - unfortunately we don't have v1.4.0 for our architecture, so I cannot compare that.

**Comment by dblate** [comment]
@mgiessing [mgiessing](https://github.com/mgiessing)
So do you solve you problem? How do you use tensorboard in pipeline now.
I get the same issue with your first option, I just can not open tensorboard until the component is completed.

**Comment by kujyp** [comment]
> 2. Mount a volume (maybe the NB server volume?) into the pipeline component and start the Tensorboard via the Kubeflow Dashboard Tab (the way which is also shown in the yt video). Probably sth. similar to this example I guess: https://github.com/kubeflow/pipelines/blob/master/samples/core/volume_ops/volume_ops.py
The advantage I see with the latter is that I can also start using/inspecting TensorBoard while the model is still training, however I read, that those will make the pipeline less flexible/reusable and should be avoided if possible...?

https://github.com/kubeflow/pipelines/blob/1.8.18/docs/config/volume-support.md
`volume-support` feature implemented from https://github.com/kubeflow/pipelines/pull/4236 
It can solve the tensorboard with training issues.
Attach PVC to Training task, then tensorboard can be opened with a PVC volume while training.

This feature is available from [kubeflow 1.1.2](https://github.com/kubeflow/pipelines/releases/tag/1.1.2)
But, [according to the document](https://github.com/kubeflow/pipelines/blob/1.8.18/docs/config/volume-support.md) it's still alpha.

@chensun 
`volume-support` feature is still in alpha. [Feature PR](https://github.com/kubeflow/pipelines/pull/4236) was merged 2 years ago. 
Any plans? To be removed feature? or to be released?
